staging/src/k8s.io/apiserver/pkg/storage/etcd3/etcdfeature/feature_support_checker.go

```diff
@@ -0,0 +1,147 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcdfeature
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/apiserver/pkg/storage"
+	"k8s.io/klog/v2"
+)
+
+var (
+	// Define these static versions to use for checking version of etcd, issue on kubernetes #123192
+	v3_4_25 = version.MustParseSemantic("3.4.25")
+	v3_5_0  = version.MustParseSemantic("3.5.0")
+	v3_5_8  = version.MustParseSemantic("3.5.8")
+
+	// DefaultFeatureSupportChecker is a shared global FeatureSupportChecker.
+	DefaultFeatureSupportChecker FeatureSupportChecker = newDefaultFeatureSupportChecker()
+)
+
+// FeatureSupportChecker to define Supports functions.
+type FeatureSupportChecker interface {
+	Supports(feature storage.Feature) (bool, error)
+	// This method works with etcd client to recalcuate feature support and caches it.
+	EndpointSupports(ctx context.Context, c client, endpoint string, feature storage.Feature) (bool, error)
+	// Only use for testing and use defer afterwork.
+	SetFeatureSupportCheckerDuringTest(feature storage.Feature, value bool) func()
+}
+
+// featureSupportInfo holds information about feature support status, used for cache.
+type featureSupportInfo struct {
+	Supported bool
+}
+
+type defaultFeatureSupportChecker struct {
+	progresNotifySupport sync.Map
+}
+
+func newDefaultFeatureSupportChecker() *defaultFeatureSupportChecker {
+	return &defaultFeatureSupportChecker{}
+}
+
+// Supports can check the featue from anywhere without storage if it was checked before with Supports
+// Only return true or false if it was defined or not.
+func (f *defaultFeatureSupportChecker) Supports(feature storage.Feature) (bool, error) {
+	switch feature {
+	case storage.RequestWatchProgress:
+		return f.supportsRequestWatchProgress()
+	default:
+		return false, fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", feature)
+	}
+}
+
+func (f *defaultFeatureSupportChecker) EndpointSupports(ctx context.Context, c client, endpoint string, feature storage.Feature) (bool, error) {
+	if value, ok := f.progresNotifySupport.Load(endpoint); ok {
+		return value.(featureSupportInfo).Supported, nil
+	}
+
+	supported, err := endpointSupportsRequestWatchProgress(ctx, c, endpoint)
+	if err != nil {
+		return false, err
+	}
+	f.progresNotifySupport.Store(feature, featureSupportInfo{
+		Supported: supported,
+	})
+	return supported, nil
+}
+
+func (f *defaultFeatureSupportChecker) supportsRequestWatchProgress() (bool, error) {
+	isSupported := true
```

Comment:
shouldn't this be set to `false` ?
if yes, then you could remove this variable and simply return false in line `96`.

Reply:
If there are no cached endpoints, we can assume it's supported.